server/pulp/server/lazy/url.py

```diff
@@ -0,0 +1,594 @@
+        :rtype: str
+        """
+        policy = {
+            Policy.RESOURCE: self.resource,
+            Policy.EXPIRATION: self.expiration,
+            Policy.EXTENSIONS: self.extensions
+        }
+        policy = JSON.encode(policy)
+        policy = Base64.encode(policy)
+        return policy
+
+    def sign(self, key):
+        """
+        Sign the policy using the specified private RSA key.
+
+        :param key: A private RSA key.
+        :type key: RSA.RSA
+        :return A tuple of: (encoded-policy, base64-signature)
+        :rtype tuple
+        """
+        encoded = self.encode()
+        digest = Policy.digest(encoded)
+        signature = key.sign(digest)
+        return encoded, Base64.encode(signature)
+
+    def __str__(self):
+        return str(self.__dict__)
+
+
+class Query(object):
+    """
+    URL query.
+    """
+
+    @staticmethod
+    def decode(encoded):
+        """
+        Decode the URL formatted parameters into a dictionary.
+
+        :param encoded: A URL formatted parameters.
+        :type encoded: str
+        :return: The parameters as a dictionary.
+        :rtype: dict
+        """
+        decoded = {}
+        for pair in encoded.split(';'):
+            if not pair:
+                continue
+            k, v = pair.split('=')
+            decoded[k] = unquote(v)
+        return decoded
+
+    @staticmethod
+    def encode(parsed):
+        """
+        Join the parsed parameters dictionary into URL parameter format.
+
+        :param parsed: A parsed parameters.
+        :type parsed: dict
+        :return: A parameters string.  Eg: name=john;age=10
+        :rtype: str
+        """
+        encoded = []
+        for k, v in parsed.items():
+            p = '{k}={v}'.format(k=k, v=quote(v))
+            encoded.append(p)
+        return ';'.join(encoded)
+
+
+class Key(object):
+    """
+    Provides RSA key management.
+    """
+
+    @staticmethod
+    def load(path=None, pem=None):
+        """
+        Get/Load an RSA key at the specified path.
+
+        :param path: An absolute path to a PEM encoded key.
+        :type path: str
+        :param pem: A PEM encoded key.
+        :type pem: str
+        :return: The loaded key.
+        :rtype: RSA.RSA
+        """
+        if path:
+            with open(path) as fp:
+                pem = fp.read()
+        bfr = BIO.MemoryBuffer(pem)
+        if 'PRIVATE' in pem:
+            key = RSA.load_key_bio(bfr)
+        else:
+            key = RSA.load_pub_key_bio(bfr)
+        return key
+
+
+class URL(object):
+    """
+    An URL object that supports signing.
+
+    :ivar content: The actual URL.
+    :type content: ParseResult
+    """
+
+    POLICY = 'policy'
+    SIGNATURE = 'signature'
+
+    def __init__(self, content):
+        """
+        :param content: The *actual* URL key.
+        :type content: str
+        """
+        self.content = urlparse(content)
+
+    @property
+    def scheme(self):
```

Comment:
Since it's a `URL` class, the properties provide an intuitive interface for the class.  They also provide *some* encapsulation of how the url is stored internally.  Although, making `content` be protected `_content` would complete this argument.